Fix clone screen issues with gtk plugin on xinerama displays

From: Thorsten Behrens <thb@openoffice.org>


---

 vcl/unx/gtk/app/gtkdata.cxx |   25 +++++++++++++++++++++++--
 1 files changed, 23 insertions(+), 2 deletions(-)


diff --git vcl/unx/gtk/app/gtkdata.cxx vcl/unx/gtk/app/gtkdata.cxx
index 047938a..2ebd8b6 100644
--- vcl/unx/gtk/app/gtkdata.cxx
+++ vcl/unx/gtk/app/gtkdata.cxx
@@ -224,8 +224,29 @@ void GtkSalDisplay::monitorsChanged( GdkScreen* pScreen )
                 {
                     GdkRectangle dest;
                     gdk_screen_get_monitor_geometry(pScreen, i, &dest);
-                    m_aXineramaScreens.push_back( Rectangle( Point(dest.x,
-                                                                   dest.y ), Size( dest.width, dest.height ) ) );
+
+                    // see if any frame buffers are at the same coordinates
+                    // this can happen with weird configuration e.g. on
+                    // XFree86 and Clone displays
+                    bool bDuplicate = false;
+                    for( int n = 0; n < i; n++ )
+                    {
+                        if( m_aXineramaScreens[n].Left() == dest.x &&
+                            m_aXineramaScreens[n].Top() == dest.y )
+                        {
+                            bDuplicate = true;
+                            if( m_aXineramaScreens[n].GetWidth() < dest.width ||
+                                m_aXineramaScreens[n].GetHeight() < dest.height )
+                            {
+                                m_aXineramaScreens[n].SetSize( Size( dest.width,
+                                                                     dest.height ) );
+                            }
+                            break;
+                        }
+                    }
+                    if( ! bDuplicate )
+                        m_aXineramaScreens.push_back( Rectangle( Point(dest.x,
+                                                                       dest.y ), Size( dest.width, dest.height ) ) );
                 }
                 m_bXinerama = m_aXineramaScreens.size() > 1;
                 if( ! m_aFrames.empty() )
